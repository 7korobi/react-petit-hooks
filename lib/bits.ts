const BITLIMIT = 31
const BITMASK = 2 ** 31 - 1

type Labels<T extends readonly string[]> = T[number][]
type BitsDic<T extends readonly string[], X> = {
  [key in T[number]]: X
}

export class BitsData<T extends readonly string[]> extends Array {
  0: number
  1: BitsShadow<T>
  constructor(value: number, shadow: BitsShadow<T>) {
    super(2)
    this[0] = value
    this[1] = shadow
  }
}

export class BitsShadow<T extends readonly string[]> {
  labels: T
  mask: number
  posi: BitsDic<T, number>
  nega: BitsDic<T, number>
  idx: BitsDic<T, number>

  constructor(labels: T) {
    if (BITLIMIT < labels.length) {
      throw new Error('too much bits.')
    }
    this.labels = labels
    this.mask = 0
    this.posi = {} as any
    this.nega = {} as any
    this.idx = {} as any
    labels.forEach((label: T[number]) => {
      this.posi[label] = this.idx[label] = 0
      this.nega[label] = BITMASK
    })
    labels.forEach((label: T[number], idx) => {
      const posi = 2 ** idx
      const nega = BITMASK & ~posi
      this.mask |= posi
      this.posi[label] |= posi
      this.nega[label] &= nega
      this.idx[label] || (this.idx[label] = idx)
    })
  }

  by(src: T | Labels<T>): number
  by(src: number): Labels<T>
  by(src: T | Labels<T> | number): Labels<T> | number {
    if ('number' === typeof src) {
      const labels: Labels<T> = []
      this.labels.forEach((label: T[number]) => {
        if (src & this.posi[label]) {
          labels.push(label)
        }
      })
      return labels
    } else {
      let n = 0
      src.forEach((label: T[number]) => {
        n |= (this.posi[label] || 0) as number
      })
      return n
    }
  }

  is(n: number): [BitsData<T>, BitsDic<T, number>] {
    const o: BitsData<T> = new BitsData<T>(n, this)
    const proxy = new Proxy(o, {
      get([value, shadow], label: T[number]) {
        return value & shadow.posi[label]
      },
      set(data, label: T[number], set: number | boolean | null) {
        const [value, shadow] = data
        let bits: number
        switch (set) {
          case null:
          case false:
            bits = 0
            break
          case true:
            bits = shadow.posi[label]
            break
          default:
            bits = shadow.posi[label] & (set << shadow.idx[label])
        }
        data[0] = value & shadow.nega[label] | bits
        return true
      },
      has([value, shadow], label: T[number]) {
        return !!shadow.idx[label]
      },
    }) as any
    return [o, proxy]
  }

  isOneBit(x: number) {
    return 0 === (x & (x - 1))
  }

  firstOut(x: number) {
    return x & (x - 1)
  }
  firstIn(x: number) {
    return x | (x + 1)
  }

  firstLinksOut(x: number) {
    return ((x | (x - 1)) + 1) & x
  }
  firstLinksIn(x: number) {
    return ((x & (x + 1)) - 1) | x
  }

  findBitOn(x: number) {
    return x & -x
  }
  findBitOff(x: number) {
    return ~x & (x + 1)
  }

  fillHeadsToOn(x: number) {
    return x | (x - 1)
  }

  fillHeadsToOff(x: number) {
    return x & (x + 1)
  }

  headsBitOff(x: number) {
    return ~x & (x - 1)
  }
  headsBitOn(x: number) {
    return ~(~x | (x + 1))
  }

  headsBitOffAndNextOn(x: number) {
    return x ^ (x - 1)
  }

  snoob(x: number) {
    const minbit = x & -x
    const ripple = x + minbit
    const ones = ((x ^ ripple) >>> 2) / minbit
    return ripple | ones
  }

  count(x: number) {
    let n
    n = (x >>> 1) & 0x77777777
    x = x - n
    n = (n >>> 1) & 0x77777777
    x = x - n
    n = (n >>> 1) & 0x77777777
    x = x - n
    x = (x + (x >>> 4)) & 0x0f0f0f0f
    x = x * 0x01010101
    return x >>> 24
  }

  reverse(x: number) {

  }
}

export function Bits<T extends readonly string[]>(labels: T) {
  return new BitsShadow(labels)
}
