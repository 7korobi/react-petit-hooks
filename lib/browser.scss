body {
  overscroll-behavior: none;
  pointer-events: none;
  user-select: none;

  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

#safe-area-measure {
  pointer-events: none;
  user-select: none;
  opacity: 0;
  visibility: hidden;

  z-index: -99999;

  width: 1%;
  height: 1%;
  margin-top: env(safe-area-inset-top);
  margin-right: env(safe-area-inset-right);
  margin-bottom: env(safe-area-inset-bottom);
  margin-left: env(safe-area-inset-left);
}