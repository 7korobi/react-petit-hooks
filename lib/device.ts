export const __SPEC__ = 'undefined' === typeof window
export const __BROWSER__ = !__SPEC__
export const __TEST__ = __SPEC__ || 'localhost' === location.hostname
