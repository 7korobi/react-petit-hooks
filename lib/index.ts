export * from './browser'
export * from './poll'
export * from './storage'
export * from './time'

