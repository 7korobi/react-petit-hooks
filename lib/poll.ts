import { useState, useEffect, Dispatch, SetStateAction } from 'react'
import Dexie from 'dexie'

import { to_tempo } from './time'
import { useInternet, useVisible } from './browser'

type DIC<T> = { [idx: string]: T }
type API<T> = (...args: any[]) => Promise<T>

const roops: DIC<() => Promise<void>> = {}
const timers: DIC<NodeJS.Timeout> = {}
const is_cache: DIC<number> = {}

const dexie = new Dexie('poll-web')
dexie.version(1).stores({
  meta: '&idx',
  data: '&idx',
})

export function usePoll<T>(
  api: API<T>,
  initState: T,
  timestr: string,
  version: string,
  args: any[] = []
): [T, Dispatch<SetStateAction<T>>] {
  const [list, setList] = useState(initState)

  const [is_online] = useInternet()
  const [is_visible] = useVisible()

  const is_active = is_online && is_visible
  const idx = [api.name, ...args].join('&')
  let tempo = to_tempo(timestr)

  useEffect(() => {
    if (is_active) {
      Object.values(roops).forEach((roop) => roop())
    } else {
      Object.values(timers).forEach(clearTimeout)
    }
  }, [is_active])

  useEffect(() => {
    roops[idx] = roop
    const p = roop()
    return () => {
      clearTimeout(timers[idx])
      delete roops[idx]
      delete timers[idx]
    }
  }, [])
  return [list, setList]

  async function roop() {
    let meta: { idx: string; version: string; next_at: number } | null = null
    let data: { idx: string; pack: any } | null = null
    tempo = tempo.reset()
    const { timeout, write_at, next_at } = tempo
    try {
      if (write_at < is_cache[idx]) {
        get_pass()
      } else {
        // IndexedDB metadata not use if memory has past data,
        if (!(0 < is_cache[idx])) {
          meta = await dexie.table('meta').get(idx)
          if (meta!?.version !== version) {
            meta = null
          }
        }
        if (write_at < meta!?.next_at) {
          await get_by_lf()
        } else if (0 < meta!?.next_at) {
          await get_by_lf()
          await get_by_api()
        } else {
          await get_by_api()
          dexie.table('meta').put({ idx, version, next_at })
        }
      }
      is_cache[idx] = next_at
    } catch (e) {
      console.error(e)
    }
    if (timeout < 0x7fffffff) {
      // 25days
      timers[idx] = (setTimeout(roop, timeout) as unknown) as NodeJS.Timeout
    }

    async function get_pass() {
      console.log({ wait: new Date().getTime() - write_at, idx, mode: null })
    }
    async function get_by_lf() {
      data = await dexie.table('data').get(idx)
      // Mem.State.store(meta)
      setList(data!?.pack)
      console.log({ wait: new Date().getTime() - write_at, idx, mode: '(lf)' })
    }
    async function get_by_api() {
      const pack = await api(...args)
      // TODO: data to meta api.
      // meta = Mem.State.transaction(()=> cb(data),{})
      data = { idx, pack }
      await dexie.table('data').put(data)
      setList(data!?.pack)
      console.log({
        wait: new Date().getTime() - write_at,
        idx,
        mode: '(api)',
      })
    }
  }
}
