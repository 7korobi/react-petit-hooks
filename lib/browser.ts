import { useEffect, useState, useReducer } from "react"
import { __BROWSER__ } from "./device"


declare global {
  interface Window {
    visualViewport: {
      offsetLeft: number
      offsetTop: number
      pageLeft: number
      pageTop: number
      width: number
      height: number
      scale: number
      addEventListener(type: "resize" | "scroll", cb: (e: Event) => void): void
      removeEventListener(
        type: "resize" | "scroll",
        cb: (e: Event) => void
      ): void
    }
  }
}

type SIZE = [number, number]

const vp = __BROWSER__
  ? window.visualViewport
  : { width: 0, height: 0, scale: 1 }

export const ViewBox: { size: SIZE; scale: number } = {
  size: [vp.width, vp.height],
  scale: vp.scale,
}

export function useInternet(): [boolean] {
  const [isOnline, setIsOnline] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      window.addEventListener("offline", network_state)
      window.addEventListener("online", network_state)
      return () => {
        window.removeEventListener("offline", network_state)
        window.removeEventListener("online", network_state)
      }
    }, [])
  }
  return [isOnline]

  function network_state() {
    setIsOnline(window.navigator.onLine)
  }
}

export function useVisible(): [boolean] {
  const [isVisible, setIsVisible] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      document.addEventListener("visibilitychange", visible_state)
      return () => {
        document.removeEventListener("visibilitychange", visible_state)
      }
    }, [])
  }
  return [isVisible]

  function visible_state() {
    setIsVisible("hidden" !== document.visibilityState)
  }
}

export function useKeyboard(): [] {
  if (__BROWSER__) {
    useEffect(() => {
      document.addEventListener("keyup", keyboard)
      return () => {
        document.removeEventListener("keyup", keyboard)
      }
    }, [])
  }

  return []
  function keyboard(e: KeyboardEvent) {
    console.log(e)
  }
}

export function useContextMenu(): [boolean, (isMenu: boolean) => void] {
  const [isMenu, setIsMenu] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      const { style } = document.body
      if (isMenu) {
        style.setProperty("--menu-display", "block")
      } else {
        style.setProperty("--menu-display", "none")
      }
    }, [isMenu])

    useEffect(() => {
      document.addEventListener("contextmenu", deny)
      return () => {
        document.removeEventListener("contextmenu", deny)
      }
    }, [])
  }

  return [isMenu, setIsMenu]

  function deny(e: { preventDefault: () => void }) {
    e.preventDefault()
    setIsMenu(true)
  }
}

export function useViewport(): [number, number, number] {
  const [width, setWidth] = useState<number>(ViewBox.size[0])
  const [height, setHeight] = useState<number>(ViewBox.size[1])
  const [scale, setScale] = useState<number>(ViewBox.scale)

  if (__BROWSER__) {
    useEffect(() => {
      resize()
      window.visualViewport.addEventListener("resize", resize)
      window.addEventListener("orientationChange", resize)

      return () => {
        window.visualViewport.removeEventListener("resize", resize)
        window.addEventListener("orientationChange", resize)
      }
    }, [])
  }

  return [width, height, scale]

  function resize() {
    const { style } = document.body
    const { scale, width, height } = window.visualViewport
    ViewBox.scale = scale

    setScale(scale)
    if (scale === 1) {
      style.setProperty("--view-width", `${width}px`)
      style.setProperty("--view-height", `${height}px`)

      ViewBox.size = [width, height]
      setWidth(width)
      setHeight(height)
    }
  }
}

interface Axis {
  label(n: number | null, oldVal?: string): string | null
}

class AccelAxis implements Axis {
  constructor(
    public min: number,
    public max: number,
    public minus: string,
    public plus: string
  ) {}
  label(n: number | null, oldVal: string) {
    if (null === n) {
      return null
    }
    if (n < -4) {
      return this.minus
    }
    if (4 < n) {
      return this.plus
    }
    if (oldVal === this.minus && n < -1) {
      return this.minus
    }
    if (oldVal === this.plus && 1 < n) {
      return this.plus
    }
    return ""
  }
}
class RotateAxis implements Axis {
  constructor(
    public min: number,
    public max: number,
    public minus: string,
    public plus: string
  ) {}
  label(n: number | null, oldVal: string) {
    if (null === n) {
      return null
    }
    if (n < -140) {
      return this.minus
    }
    if (140 < n) {
      return this.plus
    }
    if (oldVal === this.minus && n < -30) {
      return this.minus
    }
    if (oldVal === this.plus && 30 < n) {
      return this.plus
    }
    return ""
  }
}
class CompassAxis implements Axis {
  constructor(public min: number, public max: number) {}
  label(n: number | null) {
    if (null === n) {
      return null
    }
    const n1 = Math.floor(n)
    const n2 = Math.floor((n * 60) % 60)
    const n3 = Math.floor((n * 3600) % 60)

    return `${n1}°${n2}′${n3}″`
  }
}
class GeoAxis implements Axis {
  constructor(
    public min: number,
    public max: number,
    public minus: string,
    public plus: string
  ) {}
  label(n: number | null) {
    if (null === n) {
      return null
    }
    const n1 = Math.floor(n)
    const n2 = Math.floor((n * 60) % 60)
    const n3 = Math.floor((n * 3600) % 60)
    const n4 = Math.floor((n * 216000) % 60)
    const n5 = Math.floor((n * 12960000) % 60)

    const mark = n < 0 ? this.minus : this.plus
    return `${n1}°${n2}′${n3}″${n4}‴${n5}⁗${mark}`
  }
}
class MksAxis implements Axis {
  constructor(public min: number, public max: number) {}
  label(n: number | null) {
    if (null === n) {
      return null
    }
    const n1 = Math.floor(n)
    const n2 = Math.floor((n * 100) % 100)
    return `${n1}ｍ${n2}㎝`
  }
}

// 運動座標系
const xAxis = new AccelAxis(-10, 10, "右", "左") // m/s2
const yAxis = new AccelAxis(-10, 10, "上", "下") // m/s2
const zAxis = new AccelAxis(-10, 10, "表", "裏") // m/s2

const arAxis = new RotateAxis(-360, 360, "俯下", "仰上") // degree/s
const brAxis = new RotateAxis(-360, 360, "左折", "右折") // degree/s
const grAxis = new RotateAxis(-360, 360, "左巻", "右巻") // degree/s

const altAxis = new MksAxis(0, 10) // m
const spdAxis = new MksAxis(0, 10) // m/s

// 地球座標系
const aAxis = new CompassAxis(0, 360) // degree
const bAxis = new GeoAxis(-180, 180, "左折", "右折") // degree
const gAxis = new GeoAxis(-90, 90, "左巻", "右巻") // degree

const latAxis = new GeoAxis(-180, 180, "E", "W")
const lonAxis = new GeoAxis(-90, 90, "N", "S")

const headAxis = new CompassAxis(0, 360) // degree

export function useDeviceOrientation<U>(
  rotate: (oldVal: U, newVal: number | null, axis: CompassAxis | RotateAxis) => U
): [[U, U, U], boolean] {
  const [alpha, setAlpha] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, aAxis)
  )
  const [beta, setBeta] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, bAxis)
  )
  const [gamma, setGamma] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, gAxis)
  )
  const [absolute, setAbsolute] = useState(true)

  if (__BROWSER__) {
    useEffect(() => {
      if (DeviceOrientationEvent.requestPermission) {
        DeviceOrientationEvent.requestPermission().then((ps) => {
          if ("granted" === ps) {
            window.addEventListener("deviceorientation", onOrientation)
          }
        })
      } else {
        window.addEventListener("deviceorientation", onOrientation)
      }
      return () => {
        window.addEventListener("deviceorientation", onOrientation)
      }
    }, [])
  }

  return [[alpha, beta, gamma], absolute]

  function onOrientation({
    alpha,
    beta,
    gamma,
    absolute,
  }: DeviceOrientationEvent) {
    setAlpha([alpha!, aAxis])
    setBeta([beta!, bAxis])
    setGamma([gamma!, gAxis])
    setAbsolute(absolute)
  }

  function chkRotate(
    oldVal: U,
    [newVal, axis]: [number | null, CompassAxis | RotateAxis]
  ): U {
    return rotate(oldVal, newVal, axis)
  }
}

export function useDeviceMotion<T, U>(
  motion: (oldVal: T, newVal: number | null, axis: AccelAxis) => T,
  rotate: (oldVal: U, newVal: number | null, axis: RotateAxis) => U
): [[T, T, T], [T, T, T], [T, T, T], [U, U, U], number] {
  const [ax, setAx] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, xAxis)
  )
  const [ay, setAy] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, yAxis)
  )
  const [az, setAz] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, zAxis)
  )

  const [gx, setGx] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, xAxis)
  )
  const [gy, setGy] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, yAxis)
  )
  const [gz, setGz] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, zAxis)
  )

  const [mx, setMx] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, xAxis)
  )
  const [my, setMy] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, yAxis)
  )
  const [mz, setMz] = useReducer<typeof chkMotion>(
    chkMotion,
    motion(null!, null, zAxis)
  )

  const [alpha, setAlpha] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, arAxis)
  )
  const [beta, setBeta] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, brAxis)
  )
  const [gamma, setGamma] = useReducer<typeof chkRotate>(
    chkRotate,
    rotate(null!, null, grAxis)
  )
  const [interval, setInterval] = useState(0)

  if (__BROWSER__) {
    useEffect(() => {
      if (DeviceMotionEvent.requestPermission) {
        DeviceMotionEvent.requestPermission().then((ps) => {
          if ("granted" === ps) {
            window.addEventListener("devicemotion", onMotion)
          }
        })
      } else {
        window.addEventListener("devicemotion", onMotion)
      }
      return () => {
        window.addEventListener("devicemotion", onMotion)
      }
    }, [])
  }

  return [
    [ax, ay, az],
    [gx, gy, gz],
    [mx, my, mz],
    [alpha, beta, gamma],
    interval,
  ]

  function onMotion({
    interval,
    acceleration,
    accelerationIncludingGravity,
    rotationRate,
  }: DeviceMotionEvent) {
    const { x: ax, y: ay, z: az } = acceleration!
    const { x: mx, y: my, z: mz } = accelerationIncludingGravity!
    const { alpha, beta, gamma } = rotationRate!
    setAx([ax!, xAxis])
    setAy([ay!, yAxis])
    setAz([az!, zAxis])
    setMx([mx!, xAxis])
    setMy([my!, yAxis])
    setMz([mz!, zAxis])
    setGx([mx! - ax!, xAxis])
    setGy([my! - ay!, yAxis])
    setGz([mz! - az!, zAxis])
    setAlpha([alpha!, arAxis])
    setBeta([beta!, brAxis])
    setGamma([gamma!, grAxis])
    setInterval(interval)
  }

  function chkMotion(oldVal: T, [newVal, axis]: [number | null, AccelAxis]): T {
    return motion(oldVal, newVal, axis)
  }
  function chkRotate(
    oldVal: U,
    [newVal, axis]: [number | null, RotateAxis]
  ): U {
    return rotate(oldVal, newVal, axis)
  }
}

export function useGeoLocation<T, U>(
  geo: (oldVal: T, newVal: number | null, axis: GeoAxis | CompassAxis) => T,
  mks: (oldVal: U, newVal: number | null, axis: MksAxis) => U
): [[T, T], U, T, U] {
  const [latitude, setLatitude] = useReducer<typeof chkGeo>(
    chkGeo,
    geo(null!, null, latAxis)
  )
  const [longitude, setLongitude] = useReducer<typeof chkGeo>(
    chkGeo,
    geo(null!, null, lonAxis)
  )
  const [heading, setHeading] = useReducer<typeof chkCompass>(
    chkCompass,
    geo(null!, null, headAxis)
  )

  const [altitude, setAltitude] = useReducer<typeof chkMks>(
    chkMks,
    mks(null!, null, altAxis)
  )
  const [speed, setSpeed] = useReducer<typeof chkMks>(
    chkMks,
    mks(null!, null, spdAxis)
  )

  if (__BROWSER__) {
    if (!navigator?.geolocation) {
      return [[latitude, longitude], altitude, heading, speed]
    }

    useEffect(() => {
      const watch_id = navigator.geolocation.watchPosition(
        onPosition,
        ({ code }) => {
          console.log(`error watchPosition = ${code}`)
        },
        {
          enableHighAccuracy: true,
          maximumAge: 60 * 1000,
          timeout: 10 * 1000,
        }
      )
      return () => {
        navigator.geolocation.clearWatch(watch_id)
      }
    }, [])
  }

  return [[latitude, longitude], altitude, heading, speed]

  function onPosition({ coords, timestamp }: Position) {
    let { latitude, longitude, altitude, heading, speed } = coords
    setLongitude([longitude, lonAxis])
    setLatitude([latitude, latAxis])
    if (altitude !== null) setAltitude([altitude, altAxis])
    if (heading !== null) setHeading([heading, headAxis])
    if (speed !== null) setSpeed([speed, spdAxis])
  }
  function chkCompass(
    oldVal: T,
    [newVal, axis]: [number | null, CompassAxis]
  ): T {
    return geo(oldVal, newVal, axis)
  }
  function chkGeo(oldVal: T, [newVal, axis]: [number | null, GeoAxis]): T {
    return geo(oldVal, newVal, axis)
  }
  function chkMks(oldVal: U, [newVal, axis]: [number | null, MksAxis]): U {
    return mks(oldVal, newVal, axis)
  }
}
