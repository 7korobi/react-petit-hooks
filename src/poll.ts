import { useState, useEffect } from 'react'
import Dexie from 'dexie'

import { to_tempo, Tempo } from './time'
import { useMenu } from './browser'

import { __BROWSER__ } from './device'

type DATA<T> = { idx: string; pack: T }
type META = { idx: string; version: string; next_at: number }

class PollWeb extends Dexie {
  meta!: Dexie.Table<META, string>
  data!: Dexie.Table<DATA<any>, string>
}

let dexie: PollWeb = null as any
if (__BROWSER__) {
  dexie = new Dexie('poll-web') as PollWeb
  dexie.version(1).stores({
    meta: '&idx',
    data: '&idx',
  })
}

export function usePoll<T>(
  api: (...args: any[]) => Promise<{ pack: T }>,
  store: (data: DATA<T>) => void,
  initState: T | (() => T),
  timestr: string,
  version: string,
  args: any[] = []
): [T] {
  const [pack, setPack] = useState<T>(initState)
  const { isOnline, isVisible } = useMenu()
  const is_active = isOnline && isVisible
  const idx = [api.name, args.join('/')].join('&')

  // in roop vars.
  let tempo = to_tempo(timestr)
  let next_at = -Infinity
  let timerId = 0 as any

  useEffect(() => {
    if (is_active) {
      roop()
      return () => {
        clearTimeout(timerId)
      }
    } else {
      clearTimeout(timerId)
      return () => {}
    }
  }, [is_active])
  return [pack]

  async function roop() {
    tempo = tempo.reset()
    try {
      if (tempo.write_at < next_at) {
        get_pass(idx, tempo)
      } else {
        // IndexedDB metadata not use if memory has past data,
        const meta_next_at = await chk_meta(idx, version, next_at)
        if (tempo.write_at < meta_next_at) {
          await get_by_lf(idx, tempo, setPack, store)
        } else if (-Infinity < meta_next_at) {
          await get_by_lf(idx, tempo, setPack, store)
          const data = (await api(...args)) as DATA<T>
          await get_by_api(idx, tempo, setPack, data)
        } else {
          const data = (await api(...args)) as DATA<T>
          await get_by_api(idx, tempo, setPack, data)
          dexie.meta.put({ idx, version, next_at: tempo.next_at })
        }
      }
      next_at = tempo.next_at
    } catch (e) {
      console.error(e)
    }
    if (tempo.timeout < 0x7fffffff) {
      // 25days
      timerId = setTimeout(roop, tempo.timeout)
    }
  }
}

async function chk_meta(idx: string, version: string, next_at: number) {
  if (-Infinity < next_at) {
    return next_at
  }
  if (dexie) {
    const meta = await dexie.meta.get(idx)
    if (meta && meta.version === version) {
      return meta.next_at
    }
  }
  return -Infinity
}

async function get_pass(idx: string, { write_at }: Tempo) {
  const wait = new Date().getTime() - write_at
  console.log({ wait, idx, mode: null })
}

async function get_by_lf<T>(
  idx: string,
  { write_at }: Tempo,
  setPack: (pack: T) => void,
  store: (data: DATA<T>) => void
) {
  if (dexie) {
    const data = await dexie.data.get(idx)
    if (data) {
      store(data)
      setPack(data.pack)
    }
  }
  const wait = new Date().getTime() - write_at
  console.log({ wait, idx, mode: '(lf)' })
}

async function get_by_api<T>(
  idx: string,
  { write_at }: Tempo,
  setPack: (pack: T) => void,
  data: DATA<T>
) {
  if (dexie) {
    data.idx = idx
    await dexie.data.put(data)
    setPack(data.pack)
  }
  const wait = new Date().getTime() - write_at
  console.log({ wait, idx, mode: '(api)' })
}
