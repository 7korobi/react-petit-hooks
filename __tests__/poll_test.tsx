import { render, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import React from 'react'

import db from 'fake-indexeddb'
import Dexie from 'dexie'
Dexie.dependencies.indexedDB = db

import { usePoll } from '../lib/poll';

/*

interface Plan {
  _id: string,
  link: string,
  title: string,
  write_at: string,
  name: string,
  state: string,
  chip: string,
  sign: string,
  card: string[],
  upd: {
    description: string,
    time: string,
    interval: string,
    prologue: string,
    start: string,
  },
  lock: string[],
  flavor: string[],
  options: string[],
  tags: [string, string][],
}

async function api_base<T,R>(res: Response, cb: (data: T)=> { pack: R }): Promise<{ pack: R }> {
  const data: T = await res.json()
  return Mem.State.transaction(()=>{
    cb(data)
  },{})
}

async function api2(): Promise<{ pack: Plan[] }> {
  return api_base(await fetch('https://giji-api.duckdns.org/api/plan/progress'), ({plans}) => {
    Mem.Set.plans.merge(plans)
  })
}


poll = (cb)->
  methods:
    get_by_network: ->
      for key, val of @timers
        clearTimeout val

      list = cb.call @
      list.map ([name, id])=>
        idx = [name, id].join("&")
        dexie.meta.delete idx
        dexie.data.delete idx
        is_cache[idx] = 0
      @_waitwake()
 */

function delay(timeout: number){
  return new Promise((ok, ng)=>{
    setTimeout(()=>{
      try {
        ok()
      } catch (e) {
        ng(e)
      }
    },timeout)
  })
}

test('basic value', (done)=>{
  function TestData(){
    async function api1(): Promise<any[]> {
      return new Promise((ok, ng)=>{
        setTimeout(async ()=>{
          ok([{}])
          await delay(100)
          done()
        },100)
      })
    }
    const useApi1 = usePoll(api1, [], '10m', '1.0.0')
    return(<p>{useApi1[0].length ? '〇' : '✖' }</p>)
  }
  const c = render(<TestData/>)
  expect(c.container).toMatchSnapshot()
});

