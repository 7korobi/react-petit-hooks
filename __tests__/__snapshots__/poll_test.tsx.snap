// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic value 1`] = `
<div>
  <p>
    âœ–
  </p>
</div>
`;
